### PR TITLE
[WIP] Multi color text manager

### LINKED ISSUE #2019: Allow text coloring and add edge to text
## 🚀 Feature
- [ ] Ability to pass color array to text (to color each text label differently)
- [ ] Ability to set an edge (color and thickness) to text

## Motivation

We'd like to plot text color according to some values. Also, and edge around the text would be useful

## Pitch

We are working toward interfacing anndata https://github.com/theislab/anndata with napari for sptial transcriptomics data, and would like to get text visualized like this (color correspond to cluster coloring):
image

## Alternatives
For now, it only uses the first color of the array, if an array is passed to text:
```python
_text_color
# array([[0.        , 0.        , 0.        ],
#       [0.        , 0.        , 0.        ],
#       [0.        , 0.        , 0.        ],
#       ...,
#       [0.        , 0.        , 0.        ],
#       [0.15294118, 0.61960784, 0.40784314],
#       [0.        , 0.        , 0.        ]])
text = {
	"text": "{clusters}",
	"size": 24,
	"color": _text_color,  
	"anchor": "center",
	"blending": "translucent",
}
```
image

## Additional context

There is a quick fix wrt coloring (see PR) but would be happy to have inputs on whether it would be possible to also have an edge for the text

pinging @michalk8 with whom we are working on this.

Thank you!

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description
This allows text color to either be specified as a ColorManager, or as anything that can be coerced into ColorManager.colors (e.g. 'red', ['red', 'green'], [[1., 0., 0.], [0., 1., 0.]]), to support passing multiple colors as described in #2019. Here's a simple/minimal example.

```python
import numpy as np
import napari

viewer = napari.view_image(np.zeros((400, 400)))
points = np.array([[100, 100], [200, 300], [333, 111]])
properties = {
    'number': [1, 2, 3],
}
text = {
    'text': '{number}',
    'size': 20,
    'color': ['green', 'red', 'blue'],
}

points_layer = viewer.add_points(
    points,
    properties=properties,
    size=20,
    text=text,
)

napari.run()
```

Screen Shot 2021-07-01 at 2 46 45 PM

This is a WIP because I'm unsure of how TextMode and ColorMode should affect this behavior and if there are other usages I've missed (e.g. setting multiple text colors through properties). Currently, I've done something simple, which is to generate colors for text when needed and truncate/cycle ColorManager.colors to match the number of text values. But I suspect that's missing some intended use cases and maybe not taking advantage of ColorMode, so feedback on that would be really helpful!

## Type of change
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

This adds new functionality, but also breaks anyone relying on TextManager.color being a length-4 numpy array. I think it's fairly easy to make this a non-breaking change, so let me know if that's desired.

# References
This partly closes #2019, though does not cover adding borders around text, which should probably handled separately and another issue should be defined.

# How has this been tested?
- [ ] Newly added tests that cover the new functionality pass.
- [ ] All other affected tests pass.

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Review by sofroniewn — APPROVED** [review]
> Here's a simple/minimal example.

Can you add this an a simple example `python examples/multi_color_text.py`, i think will be nice to have 

I just looked through this, all looks good to me. I think we can merge now without adding color by properties and that can come later in a followup PR. But let me know @andy-sweet if you want to do it now too

@jni @kevinyamauchi feel free to review before merge too!!

**Review by kevinyamauchi — COMMENTED** [review]
Thanks for the PR, @andy-sweet ! I took it for a quick spin and it looks like the direct setting of colors is working. It is nice to be able to set colors of individual text elements! I have a couple of questions/comments about the expected behavior.

### Storing color values
In other places, the color values for all items (e.g., points, shapes) are stored on the model. For example, when a user provides a single color to be broadcasted to all Points, upon setting the color, the full array of colors is generated and stored. Here, the broadcasting/cycling to create the array of colors is performed when the `view_colors()` method is called. This is how the original `TextManager` was (i.e., it just stored the single color and vispy expanded the single color to the size of the in view text), so I don't think this is "wrong", but I wanted to point it out in case we want to take this opportunity to unify the behavior.

### Broadcasting/cycling color values
Currently, if one provides an array of colors with length not equal to the number of text, the array is resized (i.e., colors are cycled until all text elements have a color). In our other color usages, we attempt to broadcast and if that fails, we raise an error. I prefer this because if somebody doesn't know their colors are cycling, it can lead to confusing results. For example, if they accidentally provide 99 colors for 100 text elements, it might not be obvious the colors are wrong.

We could potentially use a [root validator](https://pydantic-docs.helpmanual.io/usage/validators/#root-validators) to get the number of text elements and then make a `ColorManager` with the broadcasted  colors. Alternatively, if we don't want to store the array of colors, we could test for "broadcastability" in the `TextManager.colors` validator ([already-validated values can be passed to validators as 'values'](https://pydantic-docs.helpmanual.io/usage/validators/)) What do you all think?

### Public ColorManager
Currently, the `ColorManager` instance is public at `layer.text.color`. We might want to make this private since a lot of the fields shouldn't really be modified by the user - especially if setting colors by properties is not yet supported. In other usages, the getter for the property containing the `ColorManager` returns the array of colors. Thoughts?

**Comment by andy-sweet** [comment]
Thanks for the reviews!

> Can you add this an a simple example `python examples/multi_color_text.py`, i think will be nice to have

@sofroniewn: I actually modified the existing `add_shapes_with_text.py` example because a new one would be almost exactly the same as that anyway, and the use of multiple/matching text colors makes sense there. I think that example has a bug where the order of shape edge colors is permuted (as I demoed in the dev meeting yesterday), which can hopefully be fixed by updating the `Shapes` layer to use `ColorManager` like `Points` (which doesn't have the same bug). I'll create a separate issue and may fix that in a separate PR first.

@kevinyamauchi: some really useful points and questions. Let me try to address them.

> Currently, the `ColorManager` instance is public at `layer.text.color`. We might want to make this private since a lot of the fields shouldn't really be modified by the user - especially if setting colors by properties is not yet supported.

Agreed. But I'm not sure how to do this with Pydantic in a consistent way. I read the [Pydantic docs on private attributes](https://pydantic-docs.helpmanual.io/usage/models/#private-model-attributes) and I think our `EventedModel` means that any attributes with a leading `_` will be private. But I can't seem to add a validator for the private attribute and I'm not sure if it would be serialized. Any help or existing examples in napari would be helpful!

> In other places, the color values for all items (e.g., points, shapes) are stored on the model. For example, when a user provides a single color to be broadcasted to all Points, upon setting the color, the full array of colors is generated and stored. Here, the broadcasting/cycling to create the array of colors is performed when the `view_colors()` method is called.

The main reason I cycled in `view_colors()` was because it was the simplest thing I could think of and that didn't require maintaining state across `TextManager` and `ColorManager`. As consistency is our long term goal, I don't think we should make things less consistent, so I'm going to try your suggestions. I think making them work well might require changing `TextManager.__init__` and `TextManager._set_text` but that's probably a good idea anyway.

**Comment by kevinyamauchi** [comment]
Hey @andy-sweet ! 

> Agreed. But I'm not sure how to do this with Pydantic in a consistent way. I read the Pydantic docs on private attributes and I think our EventedModel means that any attributes with a leading _ will be private. But I can't seem to add a validator for the private attribute and I'm not sure if it would be serialized. Any help or existing examples in napari would be helpful!

My recollection is that you are right about private attributes not being serialized. When I wrote that comment, I had thought that we could add a getter method to fields, but now I can't figure out a way. I think I may have been confused with the `EventedDataclass` that we experimented with before @tlambert03 made the `EventedModel`.

In an ideal world, I think the user would be able to interact with the `ColorManager`. As we discussed last week, I think this is a little tricky because currently, the `ColorManager` needs to know about the layer properties to properly set colors as a function of the properties. Perhaps this means we need to rethink the implementation.

I think an alternate approach would be to skip the `ColorManager` for now if we aren't going to add setting by properties and just convert the color argument to a RGBA array and broadcast as necessary (use the number of text elements. I think this could be done either in the color validator (if the text is validated first) or using a root validator. I am happy to expand more if it's helpful (just a bit short on time at the moment)!

**Comment by andy-sweet** [comment]
Mostly for my own record, here's my attempt to turn `color` into a private `_color` Pydantic field, but still capture a `color` input parameter in the constructor, and a public property `color` that exposes an array of colors.
- Define a field with the name `_color`.
- Add a keyword argument `color` to `__init__` with a default value of `None`, and assign it to `self._color` if it's not `None`.
- Update the Pydantic validator's decorator to use the `_color` field.
- Add a `color` property with a setter that returns `self._color.colors`.

But this gives the following error.

```
../text_manager.py:16: in <module>
    class TextManager(EventedModel):
../../../utils/events/evented_model.py:79: in __new__
    cls = super().__new__(mcs, name, bases, namespace, **kwargs)
pydantic/main.py:340: in pydantic.main.ModelMetaclass.__new__
    ???
pydantic/class_validators.py:173: in pydantic.class_validators.ValidatorGroup.check_for_unused
    ???
E   pydantic.errors.ConfigError: Validators defined with incorrect fields: _check_color (use check_fields=False if you're inheriting from the model and intended this)
```

I tried adding `check_fields=False` even though I'm not sure that's what we want, but then the color property setter isn't recognized.

```
>       text_manager_2.color = 'blue'

test_text_manager.py:128: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../utils/events/evented_model.py:142: in __setattr__
    super().__setattr__(name, value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: "TextManager" object has no field "color"
```

Based on that error message I assume you can't have property setters when extending Pydantic models. I can do the following to workaround that

```python
    def __setattr__(self, name, value):
        if name != 'color':
            super().__setattr__(name, value)
```

but at that point, understanding the implementation becomes difficult and we've drifted pretty far from the intended usage of Pydantic as I understand it. Even with that change, tests still fail because I think equality checks rely on an `==` definition from public Pydantic fields (which I think is also true for serialization).

Any other advice/tips from more experienced Pydants would be appreciated, but I'm considering making `color` a broadcastable array as suggested by @kevinyamauchi until we think that `ColorManager` is ready to be publicly exposed.

**Comment by tlambert03** [comment]
I'm sorry I haven't fully followed this conversation, what's the challenge/goal here?  do we want `.color` to be a part of the model, or a private implementation detail?  do you want it serialized?  (if you want it private except for serialization, you can probably reimplement `.dict()` and add it?)

**Comment by andy-sweet** [comment]
In my mind, these are the goals.

- `_color` be a private field of the model that gets serialized
- `color` be a property that exposes some simpler state of `_color` (e.g. a color array)

Let me try overriding `.dict()`.

**Comment by tlambert03** [comment]
gotcha, then yeah, I'd try to make `_color` a private class attribute, as you have, with a `@property` used on access..., and then for `dict` do something like:
```py
def dict(self, *args, **kwargs):
    d = super().dict(*args, **kwargs)
    d['color'] = whatever_you_want
    return d

**Comment by andy-sweet** [comment]
Sorry for the long delay on this issue. I took a break from it to work on the larger properties refactor and to get some more context on this.

@tlambert03 : thanks for the tip. I tried that out, but ultimately I still think that the usage I'm describing drifts too far from the intended usage of Pydantic and ultimately just makes things confusing. Here's a trivial example that hopefully demonstrates that: https://gist.github.com/andy-sweet/e10550283c00f319fab4bab005b1afea

Even if we decide that looks OK, I think the approach I described still has issues because whenever `color` is modified, we need access to `properties` to properly update the `ColorManager` (as @kevinyamauchi mentioned). 

At least I did learn one thing, which is that private fields do not exist in Pydantic. You can declare a private Python attribute, in which case Pydantic will not recognize that as a field, and thus will not try to validate it.

@kevinyamauchi: I tried turning `TextManager.color` into a numpy array as suggested (broadcasting colors to match the length of text values in a root validator), which almost worked. That approach breaks when `TextManager.add/remove` are called because we have to add/remove colors to keep things in sync, so the root validator will issue a warning after the first field (either color or values) is changed.

Instead of a root validator, I can add the extra `values` argument to the color validator and always update color after values in the implementation of `TextManager`. But even then I need to block the `values` event because it triggers a redraw when the sizes of `values` and `colors` don't match, and then manually emit it after both are updated. But then there are new problems like what happens when `TextManager.set_text` is called.

To summarize, I've turned this seemingly small issue into a bit of a monster, but I am struggling to see how enable this feature without defining new/inconsistent behavior (i.e. truncating/cycling/repeating colors to match the length of `values` when `TextManager.view_text` is called). @kevinyamauchi: if you have some other ideas, it might be best if we find some time to do some pair programming.

**Comment by tlambert03** [comment]
>  I still think that the usage I'm describing drifts too far from the intended usage of Pydantic and ultimately just makes things confusing. Here's a trivial example that hopefully demonstrates that

I see, so you not only want to provide some dynamic value when accessing the value, you also want to do it when setting the value (which requires a custom __setattr__)?  If so, then yes I agree, there's some model abuse still going on here.  Can we maybe sync up on this in person sometime?  It's still feeling overly complicated to me, but I know that I'm probably missing something.

For the example you gave in your test, it seemed like the main thing you wanted was the ability to set `name` on `CustomModel` with a simple string, while still having it be a complex `NameManager` object.  If that's the case, you need to tell `CustomModel` how to parse a simple string into an object using a `pre` validator:

```python
class CustomModel(EventedModel):

    name: NameManager = 'kermit'

    @validator('name', pre=True)
    def _fix_name(cls, value):
        if isinstance(value, str):
            return {'name': value}
        return value
```

**Comment by andy-sweet** [comment]
Thanks for the comments here. I think PR #3327 represents a better solution to this issue, so I'm going to close this PR to clean things up.
